### ISSUE TITLE
`PersesAtomMapper` combined with minimal redundant networks can yield more than one edge/mapping connecting two ligands in a `LigandNetwork`

### ISSUE THREAD

**Issue by dotsdl** [OpenFreeEnergy/openfe#882]
The `PersesAtomMapper` produces one or more atom mappings for any pair of ligands (`SmallMoleculeComponent`s) given to it. This appears to be expected behavior, but when combined with use of the `generate_minimal_redundant_network` planning approach, it is currently possible that the resulting `LigandNetwork` features more than one edge/mapping between any two ligands. This can complicate downstream analysis of resulting `AlchemicalNetwork`s and their results.

Is this expected behavior? It's not clear to us if this something we should be concerned about, if it's still scientifically valid, etc.

We observed this in a set of ligands used for project support within ASAP Discovery, and it broke some of our assumptions around the topology of our resulting `AlchemicalNetwork`s.

CC: @hmacdope, @jthorton

**Comment by IAlibay** [comment]
The way I understand this, there are two questions being asked here:

1. Should networks have multiple edges going between two nodes?
2. Should the output of `generate_minimal_redundant_network` creating networks which allows for multiple edges between two nodes or should it preferentially try to have multiple edges to different nodes?

Question 1:

Yes. My take is that it makes sense for networks to be allowed to have multiple edges between two nodes. I believe this was an intentional design strategy because it would allow for fun alchemistry / testing concurrent mappings / whatever other crazy things we can come up with.

Question 2:

That's an interesting one. So the way the code was written effectively just finds recursive MSTs by exclusion. This does mean that there are no protections in place to having an alternate mapping between two nodes being picked up at each MST round. 

Practically I support the idea that it should be "node redundant" not "edge redundant" - this is what I believe to be the idea behind: https://github.com/OpenFreeEnergy/konnektor/blob/main/src/konnektor/network_planners/_networkx_implementations/n_nodes_edges_network_algorithm.py

That being said, even in that algorithm I'm not convinced that we properly account for the "multiple edges between two nodes" problem. I'll ping @RiesBen here.

**Comment by jthorton** [comment]
For question 2 we could add a unique flag to the method which collapses multiple mappings between two nodes down to the single best-scoring one, this makes sense in production as we would only want to run the best possible mapping to ensure success.

**Comment by IAlibay** [comment]
> For question 2 we could add a unique flag to the method which collapses multiple mappings between two nodes down to the single best-scoring one, this makes sense in production as we would only want to run the best possible mapping to ensure success.

Yeah this + some kind of two minor versions window for changing the default behaviour would be the "safest" approach - although I could also be convinced to support an absolute behaviour change

**Comment by JenkeScheen** [comment]
Thanks for this issue @dotsdl ! I implemented the `generate_minimal_redundant_network` and failed to realize this could happen. IMO to solve question 2 (@IAlibay) we should implement some guardrails for this not to happen in that method. As Irfan says, it's important to *offer the possibility of a forward + reverse edge* between two nodes (very handy for debugging!) but for a FECs production scenario it's not necessarily useful. 

It should be straightforward to implement that guardrail. @IAlibay shall I tackle that? Or @RiesBen ?

**Comment by IAlibay** [comment]
@JenkeScheen - if you want to add the option to remove duplicate edges to the existing openfe `generate_minimal_redundant_network`, I would welcome the PR. @RiesBen will likely fix this later in Konnektor, however that's not going to get into OpenFE for a little while.
